### PR TITLE
[Security] Get Alias API wildcard exclusion with Security

### PR DESCRIPTION
**Pull request by albertzaharovits**
This builds on https://github.com/elastic/elasticsearch/pull/33518 .
Wildcard index expressions are tricky because when security is enabled,
widcards (inclusions and exclusions) have to be expanded **in the scope
of the authenticated user**. 

Relates: https://github.com/elastic/elasticsearch/issues/33805

**EDIT:**
The security plugin authorizes actions on indices. This implies that a
request with the [indices expression](https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-index.html) containing wildcards has to be
first evaluated, in the authorization layer, before the request is
handled. For authorization purposes, wildcards in expressions will
only be expanded to indices visible by the authenticated user.
However, this "constrained" evaluation has to be compatible with
the expression evaluation that a cluster without the security plugin
installed would do. Therefore every change in the evaluation logic
in any of the to sites has to be mirrored in the other.

#33518 added support forwildcard exclusion (`-...*`) at the core logic
site. This PR mirrors that at the `IndicesAndAliasesResolver` site,
which is the site that does the affore mentioned "constrained" expression
evaluation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java [hunk 1]
```diff
@@ -89,56 +83,59 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
+                                break;
                             }
                         }
                     }
+                    if (j == aliases.length) {
+                        // explicitly requested alias not excluded by any "-" wildcard in expression
+                        if (false == returnedAliasNames.contains(aliases[i])) {
+                            missingAliases.add(aliases[i]);
+                        }
+                    }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: I am not following what changes were made here, can you maybe explain? I was expecting changes needed only in IndicesAndAliasesResolver to add support for pattern exclusion, why do we need to make changes to the transport action too?

**Review by javanna — COMMENTED** [review]
I left some questions

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java [hunk 2]
```diff
@@ -77,9 +74,6 @@ public String getName() {
 
     @Override
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
-        // The TransportGetAliasesAction was improved do the same post processing as is happening here.
-        // We can't remove this logic yet to support mixed clusters. We should be able to remove this logic here
-        // in when 8.0 becomes the new version in the master branch.
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: It is not clear to me what post processing is handled here as well as in the transport action.

In the transport action, `TransportGetAliasesAction#postProcess` handles the cases when the get aliases request is of the form `/{index}/_alias`, when no alias is specified in the request, only index names.
While the code here deals with requests of the form `/_alias/{name}` or `/{index}/_alias/{name}`, taking care that only requested aliases are printed (`indicesToDisplay` and `namesProvided`) and any wildcard expressions which have not expanded to any aliases are called out in a 404 message.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java [hunk 3]
```diff
@@ -89,56 +83,59 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         getAliasesRequest.indicesOptions(IndicesOptions.fromRequest(request, getAliasesRequest.indicesOptions()));
         getAliasesRequest.local(request.paramAsBoolean("local", getAliasesRequest.local()));
 
-        //we may want to move this logic to TransportGetAliasesAction but it is based on the original provided aliases, which will
-        //not always be available there (they may get replaced so retrieving request.aliases is not quite the same).
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: This ties in with the other comments.

I think the RestHandler here tries to identify which of the many URl patterns have been invoked. For example, if the URI had index names, make sure those index names are printed even if they have no aliases. On the other hand, if aliases were requested (maybe alongside indices), than make sure all the requested aliases are returned and no empty indices without aliases will be returned.

I don't agree that the TransportAction should handle this types of corner cases. I think the TransportAction should return a minimal set of aliases (no extraneous placeholders) and the REST handler should package it with empty or null placeholders as well as to do any existence checks that are required to meet each individual REST spec.

> TransportGetAliasesAction but it is based on the original provided aliases, which will not always be available there (they may get replaced so retrieving request.aliases is not quite the same).

`GetAliasesRequest#originalAliases` contains the original provided aliases, but again I don't think **a single transport action can handle several different rest actions, especially when those rest actions have conflicting requirements, like for example the inclusion of indices with no aliases**.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: The comment is not so clear I agree, I believe it has to do with this: https://github.com/elastic/elasticsearch/pull/29538#discussion_r198836146 . In the linked PR we have made the result returned from transport client more similar to the one at REST, with the addition of empty entries for requested indices when no specific aliases were requested. Such thing was previously only happening at REST, now it happens in transport, but I believe it also still happens at REST for bw comp reasons. The idea behind the comment was "watch out, you may think something is redundant here, and ideally the post processing would happen in one place only, but it still happens in two places for bw comp reasons". Do you see what I mean? We should have done a better job at highlighting what piece of code is specifically redundant and we will be able to remove in the future. I am not sold on just removing the comment.

And yes, the code for this API is quite complicated and hard to follow.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: The other comment above is more about "we have duplicated logic which should be in one place, but there's a good reason". This comment is more about whether some parts of the REST action should be moved to the transport action. I generally prefer rest actions with almost no logic, but that has to do with trying to keep rest responses and transport client responses the same and have REST layer only doing the printing. With the transport client being deprecated, this is less and less important. 

The reason why this comment is there is because we have tried in the past to move all the logic to the transport action, but that did not quite work, also it made bw compatibility hard. We have added this comment to remember especially that `request.aliases` may get replaced, which is a tricky thing that can be easily missed. `originalAliases` were not there yet when the comment was written which means that it's a good time now to re-evaluate if this comment is still valid. 

In general, this API does quite unique things, I always found it weird that the REST layer had more information about the request than transport (till originalAliases were added to the request) and I had envisioned that one day maybe it would become more similar to the other API, with logic in the transport action and a very simple REST action. But I see how this comment does not add much value at this point, so I am fine with it being removed. Maybe we should re-evaluate what changes we want to make long-term to this API in a different PR though. See also #30536 .

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: what you explain sounds like a bug in the current wildcard exclusion implementation, besides the fact that security does not support pattern exclusions. Is that correct?

Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: Yes it is a bug in the current implementation. https://github.com/elastic/elasticsearch/pull/33518 should have added these bits as well.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: ok can we please split this into two PRs then? One for addressing the bug in the original implementation, and the new tests, the other one for adapting the security resolution code? That would make reviewing much easier.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: Okay, I understand the history behind it, and I will update the comments in between the code lines.

> Such thing was previously only happening at REST, now it happens in transport, but I believe it also still happens at REST for bw comp reasons.

In the REST layer I see that https://github.com/elastic/elasticsearch/blob/d25e3ef065f63fac4d4ff8d26252cb504ac12385/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java#L104 is working to remove entries that have been added in the transport. So I agree, there is an interplay here. Again, given the conflicting requirements I think it's better not to push all this logic on the transport layer (or suggest it in the comments).

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java (hunk 1 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java]: Okay, will do!

**Review by javanna — APPROVED** [review]
LGTM thanks @albertzaharovits

**Comment by albertzaharovits** [comment]
The failure
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/1235/console
looks like a fluke, I cannot reproduce it locally and I cannot fathom how the changes
herein could have such a ripple effect. I'll be pushing a master merge 🤞
